File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int f(int i, int[] a, int[] dp)
    {
        if(i<0) return 0;
        if(i==0) return a[i];
        if(dp[i]!=-1) return dp[i];
        int notpick = 0 + f(i-1, a, dp);
        int pick = a[i] + f(i-2, a, dp);
        return dp[i]= Math.max(pick, notpick);

    }
    public int rob(int[] nums) {
        int n = nums.length;
        int[]dp = new int[n+1];
        Arrays.fill(dp, -1);
        return f(n-1,nums,dp);
    }
}

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int solve(int i, vector<int>& nums, vector<int>& dp)
    {
        int n=nums.size();
        if(i>=n)
           return 0;
        dp[0]=nums[0];
        if(n<=1)
           return nums[0];
        dp[1]= max(nums[1],nums[0]);
        for(int i=2;i<n;i++)
        {
            dp[i]=max(nums[i]+dp[i-2],dp[i-1]);
        }      
        return dp[n-1];
    }
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int>dp(n+1,-1);
        return solve(0,nums,dp);
    }
};